README.md

```diff
@@ -375,96 +378,131 @@ switch (anInput) {
+Always use Kotlin's native types when available. Kotlin is JVM-compatible so **[TODO: more info]**
 
-### XML File Names
+### Type Inference
 
-View-based XML files should be prefixed with the type of view that they
-represent.
+Type inference should be preferred where possible to explicitly declared types. 
 
 __BAD:__
 
-- `login.xml`
-- `main_screen.xml`
-- `rounded_edges_button.xml`
+```kotlin
+val something: MyType = MyType()
+val meaningOfLife: Int = 42
+```
 
 __GOOD:__
 
-- `activity_login.xml`
-- `fragment_main_screen.xml`
-- `button_rounded_edges.xml`
+```kotlin
+val something = MyType()
+val meaningOfLife = 42
+```
 
-### Indentation
+### Constants vs. Variables 
+
+Constants are defined using the `val` keyword, and variables with the `var` keyword. Always use `val` instead of `var` if the value of the variable will not change.
+
+*Tip*: A good technique is to define everything using `val` and only change it to `var` if the compiler complains!
+
+### Companion Objects
+
+** TODO: A bunch of stuff about companion objects **
+
+### Optionals
+
+Declare variables and function return types as optional with `?` where a `null` value is acceptable.
+
+Use implicitly unwrapped types declared with `!!` only for instance variables that you know will be initialized before use, such as subviews that will be set up in `onCreate` for an Activity or `onCreateView` for a Fragment.
+
+When naming optional variables and parameters, avoid naming them like `optionalString` or `maybeView` since their optional-ness is already in the type declaration.
+
+When accessing an optional value, use optional chaining if the value is only accessed once or if there are many optionals in the chain:
 
-Similarly to Java, indentation should be __two characters__.
+```kotlin
+editText?.setText("foo")
+```
+
+**TODO: Update the rest of this section from Swift**
+
+Use optional binding when it's more convenient to unwrap once and perform multiple operations:
+
+```swift
+if let textContainer = self.textContainer {
+  // do many things with textContainer
+}
+```
 
-### Use Context-Specific XML Files
+For optional binding, shadow the original name when appropriate rather than using names like unwrappedView or actualLabel.
 
-Wherever possible XML resource files should be used:
+```swift
+Preferred:
 
-- Strings => `res/values/strings.xml`
-- Styles => `res/values/styles.xml`
-- Colors => `res/color/colors.xml`
-- Animations => `res/anim/`
-- Drawable => `res/drawable`
+var subview: UIView?
+var volume: Double?
 
+// later on...
+if let subview = subview, let volume = volume {
+  // do something with unwrapped subview and volume
+}
+```
+
+Not Preferred:
 
-### XML Attribute Ordering
+```swift
+var optionalSubview: UIView?
+var volume: Double?
 
-Where appropriate, XML attributes should appear in the following order:
+if let unwrappedSubview = optionalSubview {
+  if let realVolume = volume {
+    // do something with unwrappedSubview and realVolume
+  }
+}
+``` 
 
```

Comment:
I don't see a good place for it within an existing section. Maybe a new top level section after **Semicolons**?